### DISCUSSION TITLE
Managing state with lots of async atoms in medium/big project

### DISCUSSION THREAD

**Discussion by marioolofo** [pmndrs/jotai#1061] [Q&A]
EDIT:

Please, ignore all this nonsense code in favor of the answer bellow.
The question is still the same, manage lots of async atoms in large projects

-----------------------------

Hello everyone,

New Jotai user here with some doubts about the best way to scale my atoms in a medium/big project, hope you can have time to discuss with me some design decisions and why use one over another!

I'm updating a project that have session with jwt, ecommerce, admin with lots of forms and graphs, dashboard for clients with statistics and order status and so on.

To illustrate what i'm doing, let me use the session package as an example.
I have atoms for person data, balance and last login that uses atomWithStorage to keep track the last succeeded login and auto load the login email's input.

The implementation is as follows:

```ts
const sessionPersonAtom = atomWithStorage<SessionPerson | null>('sessionPerson', null);
const sessionPersonBalanceAtom = atomWithStorage<number | null>('sessionPersonBalance', null);
const sessionPersonLastLogin = atomWithStorage<string>('lastLogin', '');

export function useSessionPerson() {
    return useAtomValue(sessionPersonAtom);
}
... same thing for the other 2 atoms

export function useSessionActions() {
  const httpClient = useHttpClient();
  const sessionService = useSessionService();
  
  const setSessionPerson = useSetAtom(sessionPersonAtom);

  const login = (username: string, password: string) => sessionService.login(username, password).then((token) => {
    httpClient.setToken(token);
    return sessionService.getLoggedPerson().then((result) => {
      setSessionPerson(result);
      return true;
    });
  });
  
  const logout = ...
  const reloadBalance = ...
  
  return {
    login,
    logout,
    reloadBalance, ...
  };
}
```

With this implementation, I have the keep track of the state to know if data is loading or failed to load, so I came up with this composition (I know I can use atomWithQuery for something similar, but let's imagine this for a moment):

```ts
// utils/remoteAtom.ts

import { atom, Getter, Setter, WritableAtom } from 'jotai';
import { atomWithReducer } from 'jotai/utils';

export type RemoteValue<T> = {
	data: T | null;
	isLoading: boolean;
	error?: string;
};

type RemoteAction<T> =
	| {
			type: 'loading';
	  }
	| {
			type: 'error';
			error: Error;
	  }
	| {
			type: 'complete';
			data: T | null;
	  };

function emptyRemoteValue<T>() {
	return { data: null, isLoading: false } as RemoteValue<T>;
}

function dataRemoteValue<T>(data: T) {
	return { data, isLoading: false } as RemoteValue<T>;
}

function loadingRemoteValue<T>() {
	return { data: null, isLoading: true } as RemoteValue<T>;
}

function errorRemoteValue<T>(error: string) {
	return { data: null, isLoading: false, error } as RemoteValue<T>;
}

function setLoading<T>(): RemoteAction<T> {
	return { type: 'loading' };
}

function setData<T>(data: T): RemoteAction<T> {
	return { type: 'complete', data };
}

function setEmpty<T>(): RemoteAction<T> {
	return { type: 'complete', data: null };
}

function setError<T>(error: Error): RemoteAction<T> {
	return { type: 'error', error };
}

export const actions = {
	setLoading,
	setEmpty,
	setData,
	setError,
};

export function atomWithRemote<T>() {
	return atomWithReducer<RemoteValue<T>, RemoteAction<T>>(
		emptyRemoteValue<T>(),
		(prev, action) => {
			if (action?.type === 'loading') {
				return loadingRemoteValue();
			}
			if (action?.type === 'error') {
				return errorRemoteValue(String(action.error));
			}
			if (action?.type === 'complete') {
				return action.data ? dataRemoteValue(action.data) : emptyRemoteValue();
			}
			return prev;
		}
	);
}

export function withAsync<U, T>(
	baseAtom: WritableAtom<RemoteValue<T>, RemoteAction<T>>,
	fetch: (get: Getter, set: Setter, value: U) => Promise<T>
) {
	return atom(
		(get) => get(baseAtom),
		async (get, set, update: U) => {
			set(baseAtom, setLoading());
			try {
				const result = await fetch(get, set, update);
				set(baseAtom, setData(result));
			} catch (e) {
				set(baseAtom, setError(e));
			}
		}
	);
}
```

Now I can use something like this:

```ts
type LoginPayload = {
	username: string;
	password: string;
};

const sessionPersonAtom = atomWithRemote<SessionPerson>();

const sessionPersonLoginAtom = withAsync(
	sessionPersonAtom,
	async (_, __, { username, password }: LoginPayload) => {
		const service = getService<SessionService>(kSessionService);
		return service.login(username, password).then((token) => {
			const httpClient = getService<HttpClient>(kHttpClient);
			httpClient.setSessionToken(token);

			return service.getLoggedUser();
		});
	}
);

export function useSessionLogin() {
	return useAtom(sessionPersonLoginAtom);
}
```

Now useSessionLogin() returns an object with { isLoading, data, error } and a write function to login the user with the credentials.
This is nice because I can set the user balance to update when the session change, but I think that it's too much of a pattern for when the state grows with lots of async atoms that need to be fetched, as for example, when the person log in, I have to fetch data for statistics, orders, news, notifications, ...

What you think? How did you structured your biggest state with simple and elegant patterns? 

Thank you,

Mario

**Comment by marioolofo** [comment]
Sorry for the dumbness in my question above,
I found out that the package have a loadable util, so I removed all this nonsense and replaced for something like this:

```ts
import { atom, Getter } from 'jotai';
import { atomWithDefault } from 'jotai/utils';

export function atomWithAsync<T, U = unknown>(
	defaultValue: T,
	fetch: (get: Getter, update?: U) => Promise<T>,
	fetchOnInit?: boolean
) {
	const baseAtom = atomWithDefault((get) => {
		if (fetchOnInit) {
			return fetch(get);
		}
		return defaultValue;
	});

	const asyncAtom = atom<T, U>(
		(get) => get(baseAtom),
		async (get, set, update?: U) => {
			const result = await fetch(get, update);
			set(baseAtom, result);
		}
	);

	return asyncAtom;
}

```

This way I can create an async atom and a loadable to bypass the react suspense:

```ts
const personBalanceRefreshAtom = atomWithAsync<number, void>(
	0,
	async (get) => {
		const sessionToken = get(tokenAtom);
		if (!sessionToken) {
			return Promise.resolve(0);
		}
		return getService<SessionService>(kSessionService).getLoggedBalance();
	},
	true
);

const personBalanceAtom = loadable(personBalanceRefreshAtom);

export function usePersonBalance() {
	const value = useAtomValue(personBalanceAtom);
	const refresh = useSetAtom(personBalanceRefreshAtom);
	return { value, refresh };
}

export function usePersonBalanceValue() {
	const value = useAtomValue(personBalanceAtom);
	return value.state === 'hasData' ? value.data : null;
}
```

Still, I think that creating 2 hooks for every async atom will eventually become a trouble, maybe a better way is to simple make some utility function to retrieve the data from the loadable?

Mario

**Reply by dai-shi** [reply]
I think you can include `loadable` in `atomWithAsync`. And, if you use `null` for no data, you could include that logic too, resulting in one hook, maybe?

**Reply by marioolofo** [reply]
yeah, the null is a little thing that I'll insert into the atomWithAsync, but the loadable I'm not able to put in because it returns a read only atom (or am I missing something?)

**Reply by dai-shi** [reply]
I'm not sure if we are talking about the same thing.
My response was about tweaking your code a little bit to avoid the trouble. At least, that was my hope. (Sorry about `null` thing, it's not so important in your case.)

**Reply by marioolofo** [reply]
Thanks for your help, if I understood correctly, you're suggesting that I return the atom inside atomWithAsync() wrapped by loadable(), right?
I don't know how I could do that as the loadable returns a read only atom, and the refetch action would not be possible, but maybe I'm missing some sort of composition that make this possible (or is this a typescript type issue?).

**Reply by dai-shi** [reply]
Well, you can compose a new atom again.
```js
const baseAtom = atom(...)
const asyncAtom = atom(...)
const loadableAtom = loadable(asyncAtom)
const finalAtom = atom(
  (get) => get(loadableAtom),
  (get, set, arg) => set(asyncAtom, arg)
)
```

**Reply by marioolofo** [reply]
hmm nice :smiley: 

I need to become more familiar with this pattern, sometimes I forgot that I can compose and extend the atoms over and over again!
thank you for your time, I think that this reduces the setup to use the atoms and made me more aware of atoms composition 

Mario

**Reply by lillheaton** [reply]
Hey @dai-shi I'm trying your above solution for a async method but allowing to insert arguments.

```js
const baseAtom = atom('')
const asyncAtom = atom<string, { myInput: string }>(
  (get) => get(baseAtom),
  async (_, set, { myInput }) => {
    const result = await fetch('www.example.com/${myInput}')
    set(baseAtom, await result.text())
  },
)
const loadableAtom = loadable(asyncAtom)
const callableAtom = atom<Loadable<string>, { myInput: string }>(
  (get) => get(loadableAtom),
  (get, set, arg) => set(asyncAtom, arg),
)

const [mystring, myMethod] = useAtom(callableAtom)
myMethod({ myInput: 'foobar' })

```

Above example does not capture if the fetch fails. The loadable only works on the get methods? 
Trying to accomplish something similar as Recoil examples: https://recoiljs.org/docs/guides/asynchronous-data-queries#userecoilrefresher where he inserts arguments to the async method.

Do you have any code examples to achieve this?

**Reply by dai-shi** [reply]
@lillheaton 
It's next level. If you want to fetch in `write` and want to use `<Suspense>` or `loadable`, [async sometimes](https://jotai.org/docs/guides/async#async-sometimes) pattern should do.
```js
const baseAtom = atom('')
const asyncAtom = atom(
  (get) => get(baseAtom),
  (_, set, { myInput }) => {
    const promise = fetch('www.example.com/${myInput}').then((result) => result.text())
    set(baseAtom, promise)
  },
)
```

**Comment by zbyte64** [comment]
Coming from recoil, I think it was rather nice to write atoms with synchronous logic even if they depended on other async atoms. It meant we didn't have to check on the loading status of an atom. With loadable & "async sometimes" patterns, the atoms that consume those async dependencies will have special routines for intermediary loading values, ie they tend to check if a dependency atom value is null and return null. 

I also found linting promises so that they are awaited to be helpful for accidentally confusing an async atom for a sync atom: https://typescript-eslint.io/rules/no-floating-promises/

**Reply by dai-shi** [reply]
I think the discussion #1061 here was for v1. v2 has changed the way async atoms work.

**Reply by zbyte64** [reply]
I see, I should probably make this a separate discussion then. 

One discovery since my last post, I was checking how the tanstack integration is fairing and they mentioned "unwrap": https://github.com/jotaijs/jotai-tanstack-query/issues/21#issuecomment-1439254172

**Reply by zbyte64** [reply]
unwrap does not block, so downstream atoms must still do a check.

**Reply by dai-shi** [reply]
Another hack idea came up since then. We can use the async atom at the higher component in the tree, so that it make sure when you use the atom with `unwrap` we can assume it's resolved. (Types don't help in this case.)
